text/0074-stateful-multisig-pallet.md

```diff
@@ -0,0 +1,293 @@
+* `Execute` is an explicit extrinsic for a simpler API. It can be optimized to be executed automatically after getting enough approvals.
+* Any user can create a multisig account and they don't need to be part of it. (Alice in the diagram)
+* A proposal is any extrinsic including control extrinsics (e.g. add/remove owner, change threshold,..etc).
+* Any multisig account owner can start a proposal on behalf of the multisig account. (Bob in the diagram)
+* Any multisig account owener can execute proposal if it's approved by enough owners. (Dave in the diagram)
+
+### State Transition Functions
+
+All functions have detailed rustdoc in [PR#3300](https://github.com/paritytech/polkadot-sdk/pull/3300). Here is a brief overview of the functions:
+
+* `create_multisig` - Create a multisig account with a given threshold and initial owners. (Needs Deposit)
+* `start_proposal` - Start a multisig proposal. (Needs Deposit)
+* `approve` - Approve a multisig proposal.
+* `revoke` - Revoke a multisig approval from an existing proposal.
+* `execute_proposal` - Execute a multisig proposal. (Releases Deposit)
+* `cancel_own_proposal` - Cancel a multisig proposal started by the caller in case no other owners approved it yet. (Releases Deposit)
+
+Note: Next functions need to be called from the multisig account itself. Deposits are reserved from the multisig account as well.
+
+* `add_owner` - Add a new owner to a multisig account. (Needs Deposit)
+* `remove_owner` - Remove an owner from a multisig account. (Releases Deposit)
+* `set_threshold` - Change the threshold of a multisig account.
+* `cancel_proposal` - Cancel a multisig proposal. (Releases Deposit)
+* `delete_multisig` - Delete a multisig account. (Releases Deposit)
+
+### Storage/State
+
+* Use 2 main storage maps to store mutlisig accounts and proposals.
+
+```rust
+#[pallet::storage]
+  pub type MultisigAccount<T: Config> = StorageMap<_, Twox64Concat, T::AccountId, MultisigAccountDetails<T>>;
+
+/// The set of open multisig proposals. A proposal is uniquely identified by the multisig account and the call hash.
+/// (maybe a nonce as well in the future)
+#[pallet::storage]
+pub type PendingProposals<T: Config> = StorageDoubleMap<
+    _,
+    Twox64Concat,
+    T::AccountId, // Multisig Account
+    Blake2_128Concat,
+    T::Hash, // Call Hash
+    MultisigProposal<T>,
+>;
+```
+
+As for the values:
+
+```rust
+pub struct MultisigAccountDetails<T: Config> {
+	/// The owners of the multisig account. This is a BoundedBTreeSet to ensure faster operations (add, remove).
+	/// As well as lookups and faster set operations to ensure approvers is always a subset from owners. (e.g. in case of removal of an owner during an active proposal)
+	pub owners: BoundedBTreeSet<T::AccountId, T::MaxSignatories>,
+	/// The threshold of approvers required for the multisig account to be able to execute a call.
+	pub threshold: u32,
+	pub creator: T::AccountId,
```

Comment:
do we really need this? looks like it is for refund deposit when the multisig is destroyed. the easy solution is to refund it to the signer that triggers the removal

Reply:
Whoever created the multisig might be not in the signers list and even if they're one of the signers, anyone from the signers can dispatch/cancel the final call. That's why it's better to return it to the original creator.

Reply:
Is it common practice though for someone to deposit and another to get their deposit back? I understand the incentive part but having someone else take my deposit if I created the account seems a bit unfair. Maybe have part  of the deposit returned to the person executing the deletion process and another part to the original creator?

Reply:
the original creator have to explicitly transfer the ownership of the multisig so there is nothing unfair. the multisig itself must already worth something more than deposit so the value of deposit won’t change anything